java/src/main/java/ai/rapids/cudf/JCudfSerialization.java

```diff
@@ -0,0 +1,412 @@
+    long left = length;
+    long at = offset;
+    while (left > 0) {
+      int amountToCopy = (int) Math.min(left, tmpBuffer.length);
+      column.copyHostBufferBytes(tmpBuffer, 0, buffer, at, amountToCopy);
+      out.write(tmpBuffer, 0, amountToCopy);
+      left -= amountToCopy;
+      at += amountToCopy;
+    }
+    return padFor64bitAlignment(out, length);
+  }
+
+  private static long copyBasicData(DataOutputStream out, ColumnVector column, long rowOffset,
+                                    long numRows, byte[] arrayBuffer) throws IOException {
+    DType type = column.getType();
+    long bytesToCopy = numRows * type.sizeInBytes;
+    long srcOffset = rowOffset * type.sizeInBytes;
+    return copyAndPad(out, column, ColumnVector.BufferType.DATA, srcOffset, bytesToCopy, arrayBuffer);
+  }
+
+  private static long copyStringData(DataOutputStream out, ColumnVector column, long rowOffset,
+                                     long numRows, byte[] arrayBuffer) throws IOException {
+    long startByteOffset = column.getStartStringOffset(rowOffset);
+    long endByteOffset = column.getEndStringOffset(rowOffset + numRows - 1);
+    long bytesToCopy = endByteOffset - startByteOffset;
+    long srcOffset = startByteOffset;
+    return copyAndPad(out, column, ColumnVector.BufferType.DATA, srcOffset, bytesToCopy, arrayBuffer);
+  }
+
+  private static long copyStringOffsets(DataOutputStream out, ColumnVector column, long rowOffset,
+                                        long numRows, byte[] arrayBuffer) throws IOException {
+    // If an offset is copied over as a part of a slice the first entry may be non-zero.  This is
+    // okay because we fix them up when they are deserialized
+    long bytesToCopy = (numRows + 1) * 4;
+    long srcOffset = rowOffset * 4;
+    return copyAndPad(out, column, ColumnVector.BufferType.OFFSET, srcOffset, bytesToCopy, arrayBuffer);
+  }
+
+  private static ColumnVector[] sliceOffColumnVectors(DType[] dataTypes, long[] nullCounts,
+                                                      TimeUnit[] timeUnits, long numRows,
+                                                      DeviceMemoryBuffer combinedBuffer,
+                                                      HostMemoryBuffer combinedBufferOnHost) {
+    int numColumns = dataTypes.length;
+    long combinedBufferOffset = 0;
+    ColumnVector[] vectors = new ColumnVector[numColumns];
+    boolean tableSuccess = false;
+    try {
+      for (int column = 0; column < numColumns; column++) {
+        DType type = dataTypes[column];
+        long nullCount = nullCounts[column];
+        TimeUnit tu = timeUnits[column];
+        DeviceMemoryBuffer validity = null;
+        DeviceMemoryBuffer data = null;
+        HostMemoryBuffer offsets = null;
+        boolean columnSuccess = false;
+        try {
+          if (nullCount > 0) {
+            long len = padFor64bitAlignment(BitVectorHelper.getValidityLengthInBytes(numRows));
```

Comment:
I vaguely recall we ran into some issues when the Java code blindly assumed that all validity vectors had been allocated with 64-bit padding.  Do we really need to pad this here?  It should be padded as part of sending it over the wire.  Similar questions for all the other places here where the code is padding before slicing.

Reply:
We are padding it as we send it over the wire only, not as we read it in.  This version of the API is calculating what the final size should be when padded, so we can write out the proper length in the protocol without buffering it all in memory.  A separate version is adding in the padding by writing directly to the outputstream.